File: apps/desktop/src/main/logging.ts
/**
 * Structured logging for the Electron main process.
 *
 * One JSON object per line, correlation fields attached by `with()`, and a hard
 * rule that content never reaches a log line: no credentials, no keystrokes, no
 * clipboard, no pixels, no raw cursor coordinates, no media (SPEC.md §13.4).
 *
 * The Go control plane enforces the same rule in
 * `services/control/internal/logging`; the two lists are kept in step by the
 * tests either side.
 */

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

export const REDACTED = '[redacted]';

const LEVEL_ORDER: Record<LogLevel, number> = { debug: 10, info: 20, warn: 30, error: 40 };

const FORBIDDEN_KEYS = [
  'password',
  'passwd',
  'secret',
  'token',
  'authorization',
  'credential',
  'apikey',
  'cookie',
  'privatekey',
  'keystroke',
  'keystrokes',
  'keytext',
  'typedtext',
  'clipboard',
  'pixels',
  'frame',
  'framedata',
  'screenshot',
  'audio',
  'video',
  'cursorx',
  'cursory',
  'cursortrail',
  'turnpassword',
];

/** Whether a field name must be redacted before it is written. */
export function isForbiddenKey(key: string): boolean {
  const normalised = key.toLowerCase().replace(/[-_.]/g, '');
  return FORBIDDEN_KEYS.some((forbidden) => normalised.includes(forbidden.replace(/_/g, '')));
}

export type Fields = Record<string, unknown>;

export interface Logger {
  debug(message: string, fields?: Fields): void;
  info(message: string, fields?: Fields): void;
  warn(message: string, fields?: Fields): void;
  error(message: string, fields?: Fields): void;
  /** Returns a child logger that stamps every line with these fields. */
  with(fields: Fields): Logger;
}

export interface LoggerOptions {
  level?: LogLevel;
  /** Static fields on every line, e.g. component and session id. */
  base?: Fields;
  write?: (line: string) => void;
  now?: () => Date;
}

function redact(fields: Fields): Fields {
  const safe: Fields = {};
  for (const [key, value] of Object.entries(fields)) {
    if (isForbiddenKey(key)) {
      safe[key] = REDACTED;
    } else if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
      safe[key] = redact(value as Fields);
    } else {
      safe[key] = value;
    }
  }
  return safe;
}

export function createLogger(options: LoggerOptions = {}): Logger {
  const level = options.level ?? 'info';
  const base = options.base ?? {};
  const now = options.now ?? (() => new Date());
  const write = options.write ?? ((line: string) => process.stdout.write(`${line}\n`));

  const emit = (recordLevel: LogLevel, message: string, fields?: Fields) => {
    if (LEVEL_ORDER[recordLevel] < LEVEL_ORDER[level]) return;
    const record = {
      time: now().toISOString(),
      level: recordLevel.toUpperCase(),
      msg: message,
      ...redact(base),
      ...redact(fields ?? {}),
    };
    write(JSON.stringify(record));
  };

  return {
    debug: (message, fields) => emit('debug', message, fields),
    info: (message, fields) => emit('info', message, fields),
    warn: (message, fields) => emit('warn', message, fields),
    error: (message, fields) => emit('error', message, fields),
    with: (fields) =>
      createLogger({
        ...options,
        level,
        base: { ...base, ...fields },
      }),
  };
}

/** A random correlation id for a session, request or peer connection. */
export function newCorrelationId(): string {
  const bytes = new Uint8Array(8);
  globalThis.crypto.getRandomValues(bytes);
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, '0')).join('');
}

File: services/control/internal/logging/logging.go
// Package logging provides the control plane's structured logging baseline:
// machine-parseable output, request/session correlation, and a hard rule that
// sensitive material never reaches a log line (SPEC.md §13.4).
package logging

import (
	"context"
	"io"
	"log/slog"
	"strings"
)

// Redacted replaces the value of any attribute whose key is forbidden.
const Redacted = "[redacted]"

// forbiddenKeys are substrings of attribute keys that must never be logged.
// Layup audits *events*, never content: no credentials, no keystrokes, no
// clipboard, no pixels, no raw cursor coordinates, no media payloads.
var forbiddenKeys = []string{
	"password", "passwd", "secret", "token", "authorization", "credential",
	"apikey", "api_key", "cookie", "privatekey", "private_key",
	"keystroke", "keystrokes", "keytext", "typedtext", "clipboard",
	"pixels", "frame", "framedata", "screenshot", "audio", "video",
	"cursorx", "cursory", "cursortrail", "turnpassword",
}

// IsForbiddenKey reports whether an attribute key must be redacted.
func IsForbiddenKey(key string) bool {
	normalised := strings.ToLower(strings.NewReplacer("-", "", "_", "", ".", "").Replace(key))
	for _, forbidden := range forbiddenKeys {
		if strings.Contains(normalised, strings.ReplaceAll(forbidden, "_", "")) {
			return true
		}
	}
	return false
}

// Options configure a logger.
type Options struct {
	Level  string // debug|info|warn|error
	Format string // json|text
	Writer io.Writer
}

// New builds the process logger. Output is machine-parseable by default.
func New(opts Options) *slog.Logger {
	var level slog.Level
	switch strings.ToLower(opts.Level) {
	case "debug":
		level = slog.LevelDebug
	case "warn":
		level = slog.LevelWarn
	case "error":
		level = slog.LevelError
	default:
		level = slog.LevelInfo
	}

	handlerOpts := &slog.HandlerOptions{Level: level}
	var base slog.Handler
	if strings.ToLower(opts.Format) == "text" {
		base = slog.NewTextHandler(opts.Writer, handlerOpts)
	} else {
		base = slog.NewJSONHandler(opts.Writer, handlerOpts)
	}
	return slog.New(&redactingHandler{inner: base})
}

// redactingHandler enforces the "never log content" rule at the handler level,
// so a careless call site cannot leak by accident.
type redactingHandler struct{ inner slog.Handler }

func (h *redactingHandler) Enabled(ctx context.Context, level slog.Level) bool {
	return h.inner.Enabled(ctx, level)
}

func (h *redactingHandler) Handle(ctx context.Context, record slog.Record) error {
	safe := slog.NewRecord(record.Time, record.Level, record.Message, record.PC)
	record.Attrs(func(attr slog.Attr) bool {
		safe.AddAttrs(redactAttr(attr))
		return true
	})
	// Correlation fields travel on the context so every line in a request or
	// session can be stitched together.
	for _, attr := range fromContext(ctx) {
		safe.AddAttrs(redactAttr(attr))
	}
	return h.inner.Handle(ctx, safe)
}

func (h *redactingHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
	safe := make([]slog.Attr, 0, len(attrs))
	for _, attr := range attrs {
		safe = append(safe, redactAttr(attr))
	}
	return &redactingHandler{inner: h.inner.WithAttrs(safe)}
}

func (h *redactingHandler) WithGroup(name string) slog.Handler {
	return &redactingHandler{inner: h.inner.WithGroup(name)}
}

func redactAttr(attr slog.Attr) slog.Attr {
	if IsForbiddenKey(attr.Key) {
		return slog.String(attr.Key, Redacted)
	}
	if attr.Value.Kind() == slog.KindGroup {
		children := attr.Value.Group()
		safe := make([]any, 0, len(children))
		for _, child := range children {
			safe = append(safe, redactAttr(child))
		}
		return slog.Group(attr.Key, safe...)
	}
	return attr
}

type contextKey struct{}

// WithFields returns a context carrying correlation fields that are added to
// every log line emitted with it.
func WithFields(ctx context.Context, attrs ...slog.Attr) context.Context {
	existing := fromContext(ctx)
	combined := make([]slog.Attr, 0, len(existing)+len(attrs))
	combined = append(combined, existing...)
	combined = append(combined, attrs...)
	return context.WithValue(ctx, contextKey{}, combined)
}

func fromContext(ctx context.Context) []slog.Attr {
	if ctx == nil {
		return nil
	}
	attrs, _ := ctx.Value(contextKey{}).([]slog.Attr)
	return attrs
}
